### PR TITLE
feat: Use indexes for LogData

### PR DESCRIPTION
**Pull request by akhercha**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Feature

## What is the current behavior?

Resolves: #537

## What is the new behavior?

- Added `SerializableFelt252Dict` struct & its unit tests
- Replaced `LogData` members & functions with `SerializableFelt252Dict` & added tests
- Replaced places where `LogData` is used/created to align with those changes

## Does this introduce a breaking change?

No

### CODE REVIEW

**Comment by akhercha** [comment]
Tried multiple things but looks like it's not possible yet.

Ori answered me on the discord for one of my questions:
> it is not possible to copy or clone Dicts currently - this isn't actually possible to implement in this context.
SquashedFelt252Dict<T> will be possible to implement if we allow casting it into some sort of an array, but we are yet to provide that possiblity

Then using a `Felt252Dict` or any others data structure that involves one is not possible.

@Sk8erboi84 is also checking but this looks like a dead end for this particular implementation, maybe we can update it to be more "Cairo-compliant" but not sure. We can also keep it as is & wait for Cairo to support it.

Happy to have your thoughts! @sparqet @zarboq

**Review by Sk8erboi84 — COMMENTED** [review]
Ill take a look into that PR

**Comment by akhercha** [comment]
Ready for a review! @Sk8erboi84

File: src/utils/serializable_dict.cairo [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    fn serialize_into(ref self: SerializableFelt252Dict<T>) -> Array<felt252> {
+        let mut serialized_data: Array<felt252> = array![];
+        self.serialize(ref serialized_data);
+        serialized_data
+    }
+
+    //
+    // Serialization of an SerializableFelt252Dict
+    //
+    // An SerializableFelt252Dict is serialized as follow:
```
Comment (Sk8erboi84) [inline src/utils/serializable_dict.cairo:204]: Add this also in the book also please

Comment (akhercha) [reply src/utils/serializable_dict.cairo:204]: Added!

File: src/utils/serializable_dict.cairo [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+impl ItemPartialEq<
+    T, impl TCopy: Copy<T>, impl TPartialEq: PartialEq<T>, impl TDrop: Drop<T>
+> of PartialEq<Item<T>> {
+    fn eq(lhs: @Item<T>, rhs: @Item<T>) -> bool {
+        if lhs.is_single() && rhs.is_single() {
+            return lhs.unwrap_single() == rhs.unwrap_single();
+        } else if lhs.is_span() && rhs.is_span() {
+            return lhs.unwrap_span() == rhs.unwrap_span();
```
Comment (Sk8erboi84) [inline src/utils/serializable_dict.cairo:67]: Can we compare 2 span like that ?

**Review by Sk8erboi84 — APPROVED** [review]

**Pull request merged**